src/tests/ftest/util/apricot/apricot/test.py

```diff
@@ -1242,8 +1245,45 @@ def remove_temp_test_dir(self):
             errors.append("Error removing temporary test files")
         return errors
 
+    def dump_engines_stacks(self, message):
+        """Dump the engines ULT stacks.
+
+        Args:
+            message (str): reason for dumping the ULT stacks. Defaults to None.
+        """
+        if self.dumped_engines_stacks is False:
+            self.dumped_engines_stacks = True
+            self.log.info("%s, dumping ULT stacks", message)
+            dump_engines_stacks(self.hostlist_servers)
+
+    def report_timeout(self):
+        """Dump ULTs stacks if this test case was timed out."""
+        super().report_timeout()
+        if self.timeout is not None and self.time_elapsed > self.timeout:
+            # dump engines ULT stacks upon test timeout
+            self.dump_engines_stacks("Test has timed-out")
+
+    def fail(self, message=None):
+        """Dump engines ULT stacks upon test failure."""
+        self.dump_engines_stacks("Test has failed")
+        super().fail(message)
+
+    def error(self, message=None):
+        """Dump engines ULT stacks upon test error."""
+        self.dump_engines_stacks("Test has errored")
+        super().error(message)
+
     def tearDown(self):
         """Tear down after each test case."""
+
+        # dump engines ULT stacks upon test failure
+        # XXX check of Avocado test status during teardown is presently useless
```

Comment:
This pylint issue needs to be fixed before landing, maybe:
```suggestion
        # Check of Avocado test status during teardown is presently useless
```

Reply:
Could be a good idea to replace the XXX with a reference to a JIRA ticket describing what should be do when the avocado issue #5217 will be solved and available